packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -5647,6 +5657,122 @@ namespace Tpetra {
+    Kokkos::deep_copy(indices, this->k_gblInds1D_);
+
+    row_ptrs_type row_ptrs_beg("row_ptrs_beg", this->k_rowPtrs_.extent(0));
+    Kokkos::deep_copy(row_ptrs_beg, this->k_rowPtrs_);
+
+    const auto N = (row_ptrs_beg.extent(0) == 0 ? 0 : row_ptrs_beg.extent(0) - 1);
+    row_ptrs_type row_ptrs_end("row_ptrs_end", N);
+
+    bool refill_num_row_entries = false;
+    if (this->k_numRowEntries_.extent(0) > 0) {
+      // Case 1: Unpacked storage
+      refill_num_row_entries = true;
+      auto num_row_entries = this->k_numRowEntries_;
+      Kokkos::parallel_for("Fill end row pointers", range_policy(0, N),
+        KOKKOS_LAMBDA(const size_t i){
+          row_ptrs_end(i) = row_ptrs_beg(i) + num_row_entries(i);
+        }
+      );
+
+    } else {
+      // mfh If packed storage, don't need row_ptrs_end to be separate allocation;
+      // could just have it alias row_ptrs_beg+1.
+      // Case 2: Packed storage
+      Kokkos::parallel_for("Fill end row pointers", range_policy(0, N),
+        KOKKOS_LAMBDA(const size_t i){
+          row_ptrs_end(i) = row_ptrs_beg(i+1);
+        }
+      );
+    }
+
+    using padding_type = Kokkos::UnorderedMap<LocalOrdinal, size_t>;
+    resizeRowPtrsAndIndices<row_ptrs_type,indices_type,padding_type>(
+        row_ptrs_beg, row_ptrs_end, indices, padding);
+
+    if (refill_num_row_entries) {
+      Kokkos::parallel_for("Fill num entries", range_policy(0, N),
+        KOKKOS_LAMBDA(const size_t i){
+          this->k_numRowEntries_(i) = row_ptrs_end(i) - row_ptrs_beg(i);
+        }
+      );
+    }
+    this->k_rowPtrs_ = row_ptrs_beg;
+    this->k_gblInds1D_ = indices;
+  }
+
+  template <class LocalOrdinal, class GlobalOrdinal, class Node>
+  void
+  CrsGraph<LocalOrdinal, GlobalOrdinal, Node>::
+  resizeForIncomingData(const RowGraph<LocalOrdinal,GlobalOrdinal,Node>& source,
+                        size_t numSameIDs,
+                        const Teuchos::ArrayView<const LocalOrdinal> &permuteToLIDs,
+                        const Teuchos::ArrayView<const LocalOrdinal> &permuteFromLIDs)
+  {
+    using LO = LocalOrdinal;
+    using GO = GlobalOrdinal;
+
+    // Resize row pointers and indices to accommodate incoming data
+    const map_type& src_row_map = *(source.getRowMap());
+    using padding_type = Kokkos::UnorderedMap<LocalOrdinal, size_t>;
+    padding_type padding(numSameIDs+permuteFromLIDs.size());
+    for (LO tgtid=0; tgtid<static_cast<LO>(numSameIDs); ++tgtid) {
+      const GO srcgid = src_row_map.getGlobalElement(tgtid);
+      auto how_much_padding = source.getNumEntriesInGlobalRow(srcgid);
+      padding.insert(tgtid, how_much_padding);
+    }
+    for (LO i=0; i<permuteToLIDs.size(); ++i) {
+      const LO tgtid = permuteToLIDs[i];
+      const GO srcgid = src_row_map.getGlobalElement(permuteFromLIDs[i]);
+      auto how_much_padding = source.getNumEntriesInGlobalRow(srcgid);
+      padding.insert(tgtid, how_much_padding);
```

Comment:
Kokkos::UnorderedMap permits insert to fail, so you should check the return value and try again if it does fail. (It's not likely outside of a parallel loop, but the interface permits it.)

Reply:
On my list :)